agent/reconnectingpty/buffered.go

```diff
@@ -0,0 +1,234 @@
+			_, err = rpty.circularBuffer.Write(part)
+			if err != nil {
+				logger.Error(ctx, "write to circular buffer", slog.Error(err))
+				rpty.metrics.WithLabelValues("write_buffer").Add(1)
+			}
+			// TODO: Instead of ranging over a map, could we send the output to a
+			// channel and have each individual Attach read from that?
+			for cid, conn := range rpty.activeConns {
+				_, err = conn.Write(part)
+				if err != nil {
+					logger.Warn(ctx,
+						"error writing to active connection",
+						slog.F("connection_id", cid),
+						slog.Error(err),
+					)
+					rpty.metrics.WithLabelValues("write").Add(1)
+				}
+			}
+			rpty.state.cond.L.Unlock()
+		}
+	}()
+
+	return rpty
+}
+
+// lifecycle manages the lifecycle of the reconnecting pty.  If the context ends
+// or the reconnecting pty closes the pty will be shut down.
+func (rpty *bufferedReconnectingPTY) lifecycle(ctx context.Context, logger slog.Logger) {
+	rpty.timer = time.AfterFunc(attachTimeout, func() {
+		rpty.Close("reconnecting pty timeout")
+	})
+
+	logger.Debug(ctx, "reconnecting pty ready")
+	rpty.state.setState(StateReady, nil)
+
+	state, reasonErr := rpty.state.waitForStateOrContext(ctx, StateClosing, nil)
+	if state < StateClosing {
+		// If we have not closed yet then the context is what unblocked us (which
+		// means the agent is shutting down) so move into the closing phase.
+		rpty.Close(reasonErr.Error())
+	}
+	rpty.timer.Stop()
+
+	rpty.state.cond.L.Lock()
+	// Log these closes only for debugging since the connections or processes
+	// might have already closed on their own.
+	for _, conn := range rpty.activeConns {
+		err := conn.Close()
+		if err != nil {
+			logger.Debug(ctx, "closed conn with error", slog.Error(err))
+		}
+	}
+	// Connections get removed once they close but it is possible there is still
+	// some data that will be written before that happens so clear the map now to
+	// avoid writing to closed connections.
+	rpty.activeConns = map[string]net.Conn{}
+	rpty.state.cond.L.Unlock()
+
+	// Log close/kill only for debugging since the process might have already
+	// closed on its own.
+	err := rpty.ptty.Close()
+	if err != nil {
+		logger.Debug(ctx, "closed ptty with error", slog.Error(err))
+	}
+
+	err = rpty.process.Kill()
+	if err != nil {
+		logger.Debug(ctx, "killed process with error", slog.Error(err))
+	}
+
+	logger.Info(ctx, "closed reconnecting pty")
+	rpty.state.setState(StateDone, xerrors.Errorf("reconnecting pty closed: %w", reasonErr))
+}
+
+func (rpty *bufferedReconnectingPTY) Attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) error {
+	logger.Info(ctx, "attach to reconnecting pty")
+
+	// This will kill the heartbeat once we hit EOF or an error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Once we are ready, attach the active connection while we hold the mutex.
+	_, err := rpty.state.waitForStateOrContext(ctx, StateReady, func(state State, err error) error {
```

Comment:
It does feel odd.

Mostly I needed to write out the buffer even if the state is closed to fix a flake where the pty closes too quickly when there is a short-lived command, and then after that I check the state and return the original `err`.

Really this will only happen in tests though, I am not sure why we spawn things like `echo hello` when real-world usage is always spawning a shell.

I could just duplicate writing out the buffer outside the wait, before returning the error.

In the case of context expiring, I can skip the function call, this is only really needed for the quickly-closing case.  I will go ahead and do that now, and let me know if it would be smarter to duplicate the buffer write!

Somewhat related, I wonder sometimes if I should rename `error` to `reason`, since it is almost more like a reason for the state change.

Reply:
Or I could actually still omit the error and just have `waitForStateOrContext` return the original error if `fn` returns nil.

Reply:
It's starting to feel awkward to pass in the function because the circumstances under which you do and do not want to execute the function are hard to express.  Another option to consider: split `waitForStateOrContext` into a `waitForStateOrContextLocked` function that does the stuff assuming the caller has locked the `cond`.  Then, you can hold the lock here, call `waitForStateOrContextLocked`, then do your extra processing before releasing the lock.

Reply:
That makes sense, and feels cleaner.  Will fix.